src/blockdata/locktime.rs

```diff
@@ -0,0 +1,735 @@
+
+//! Provides type [`LockTime`] that implements the logic around nLockTime/OP_CHECKLOCKTIMEVERIFY.
+//!
+//! There are two types of lock time: lock-by-blockheight and lock-by-blocktime, distinguished by
+//! whether `LockTime < LOCKTIME_THRESHOLD`.
+//!
+
+use core::{mem, fmt};
+use core::cmp::{PartialOrd, Ordering};
+use core::convert::TryFrom;
+use core::str::FromStr;
+use core::num::ParseIntError;
+
+use crate::consensus::encode::{self, Decodable, Encodable};
+use crate::io::{self, Read, Write};
+use crate::prelude::*;
+
+/// The Threshold for deciding whether a lock time value is a height or a time (see [Bitcoin Core]).
+///
+/// `LockTime` values _below_ the threshold are interpreted as block heights, values _above_ (or
+/// equal to) the threshold are interpreted as block times (UNIX timestamp, seconds since epoch).
+///
+/// Bitcoin is able to safely use this value because a block height greater than 500,000,000 would
+/// never occur because it would represent a height in approximately 9500 years. Conversely, block
+/// times under 500,000,000 will never happen because they would represent times before 1986 which
+/// are, for obvious reasons, not useful within the Bitcoin network.
+///
+/// [Bitcoin Core]: https://github.com/bitcoin/bitcoin/blob/9ccaee1d5e2e4b79b0a7c29aadb41b97e4741332/src/script/script.h#L39
+pub const LOCK_TIME_THRESHOLD: u32 = 500_000_000;
+
+/// Packed lock time wraps a [`LockTime`] consensus value i.e., the raw `u32` used by the network.
+///
+/// This type implements an naive ordering based on the `u32`, this is _not_ a semantically correct
+/// ordering for a lock time, hence [`LockTime`] does not implement `Ord`. This type is useful if
+/// you want to use a lock time as part of a struct and wish to derive `Ord`. For all other uses,
+/// consider using [`LockTime`] directly.
+///
+/// # Examples
+/// ```
+/// # use bitcoin::{Amount, PackedLockTime, LockTime};
+/// #[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+/// struct S {
+///     lock_time: PackedLockTime,
+///     amount: Amount,
+/// }
+///
+/// let _ = S {
+///     lock_time: LockTime::from_consensus(741521).into(),
+///     amount: Amount::from_sat(10_000_000),
+/// };
+/// ```
+#[derive(Copy, Clone, PartialEq, Eq, PartialOrd, Ord, Debug, Hash)]
+#[cfg_attr(feature = "serde", derive(Serialize, Deserialize))]
+#[cfg_attr(feature = "serde", serde(crate = "actual_serde"))]
+pub struct PackedLockTime(pub u32);
+
+impl PackedLockTime {
+    /// If [`crate::Transaction::lock_time`] is set to zero it is ignored, in other words a
+    /// transaction with nLocktime==0 is able to be included immediately in any block.
+    pub const ZERO: PackedLockTime = PackedLockTime(0);
+
+    /// Returns the inner `u32`.
+    #[inline]
+    pub fn to_u32(&self) -> u32 {
+        self.0
+    }
+}
+
+impl fmt::Display for PackedLockTime {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        write!(f, "{}", LockTime::from(*self))
+    }
+}
```

Comment:
Missing potentially interesting impls: `LowerHex`, `UpperHex`, `FromStr`, `TryFrom<&str>`, `TryFrom<String>`, `Binary`, `Octal`, although the last two are unusual.

Reply:
Added all but the last two.